Make listsinceblock refuse unknown block hash

ryanofsky opened the pull request:
Change suggested by @theuni  who noticed listsinceblock would ignore invalid block hashes causing it to return a completely unfiltered list of transactions.

ryanofsky committed:
Make listsinceblock refuse unknown block hash

Change suggested by Cory Fields <cory-nospam-@coryfields.com> who noticed
listsinceblock would ignore invalid block hashes causing it to return a
completely unfiltered list of transactions.

promag commented:
With this change there is a different behaviour with these calls:
```sh
bitcoin-cli -named listsinceblock target_confirmations=1
bitcoin-cli listsinceblock "" target_confirmations=1
```
To make both consistent deal with empty string?
```cpp
    if (!request.params[0].isNull() && !request.params[0].get_str().empty()) {
```
Otherwise tested ACK.

promag commented:
Maybe we should also raise
```cpp
throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter");
```
if an argument is not a valid block hash?

theuni commented:
Concept ACK. Agree with @promag's points. Also, because a zero hash is returned for lastblock in some cases (though maybe not anymore after this?), hash.IsNull() should probably throw an error also.

jnewbery commented:
Tested ACK 3f4318c962d398ab375039aaf70beed2c4323fe4

ryanofsky commented:
Updated 3f4318c962d398ab375039aaf70beed2c4323fe4 -> 659b2061c4329472a45e913c5d45e6ab180600a3 ([pr/since.1](https://github.com/ryanofsky/bitcoin/commits/pr/since.1) -> [pr/since.2](https://github.com/ryanofsky/bitcoin/commits/pr/since.2))

Changed to accept empty string, and added tests for empty, invalid hex, and zero strings.

promag reviewed (commented):
Invalid blockhash is not the same as block not found. The first should throw RPC_INVALID_PARAMETER. So for now rename `test_invalid_blockhash` to test_block_notfound` or something similar?

ryanofsky commented:
> Invalid blockhash is not the same as block not found

I don't know what this means. Feel free to send a patch or open a pr with whatever change you would like to see.

theuni commented:
I generally agree with @promag that an invalid hash (zero, too long, too short) should throw a different error than a valid hash but missing block, but I suspect that we don't do that in most places already, so there's not much use in differentiating here.

utACK 659b2061c4329472a45e913c5d45e6ab180600a3

promag commented:
We do have partial argument validation, mostly type validation, but in some cases value validation. It can be improved in a different PR for sure, I can take that.

jnewbery commented:
Tested ACK 659b2061c4329472a45e913c5d45e6ab180600a3

Agree that invalid block hash validation for this and other RPC methods can be added in a future PR.

promag commented:
ACK 659b206.

gmaxwell commented:
utACK

ryanofsky commented:
Should this be backported?

theuni commented:
@ryanofsky I'd say so. I've been trying to come up with a scenario where this would be a regression from a user's pov, but can't come up with anything realistic.

All I've got is "maybe some software isn't prepared to cope with error/no output on a call", but that doesn't seem like an issue since everyone has to deal with warmup failures already.

MarcoFalke commented:
utACK 659b2061c4329472a45e913c5d45e6ab180600a3

sdaftuar commented:
utACK

fanquake commented:
Backported in #11550
